src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,1207 @@
+		goto out;
+
+	stripe_ud.asu_agg_entry = entry;
+	stripe_ud.asu_bit_map = bit_map;
+	stripe_ud.asu_cell_cnt = cell_cnt;
+	rc = ABT_eventual_create(sizeof(*status), &stripe_ud.asu_eventual);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto out;
+	}
+	rc = dss_ult_create(agg_process_partial_stripe_ult, &stripe_ud,
+			    DSS_ULT_EC, 0, 0, NULL);
+	if (rc)
+		goto ev_out;
+	rc = ABT_eventual_wait(stripe_ud.asu_eventual, (void **)&status);
+	if (rc != ABT_SUCCESS) {
+		rc = dss_abterr2der(rc);
+		goto ev_out;
+	}
+	if (*status != 0) {
+		rc = *status;
+		goto ev_out;
+	}
+
+ev_out:
+	ABT_eventual_free(&stripe_ud.asu_eventual);
+
+out:
+	return rc;
+}
+
+/* Process the prior stripe. Invoked when the iterator has moved to the first
+ * extent in the subsequent.
+ */
+static int
+agg_process_stripe(struct ec_agg_entry *entry)
+{
+	vos_iter_param_t	iter_param = { 0 };
+	struct vos_iter_anchors	anchors = { 0 };
+	bool			update_vos = true;
+	int			rc = 0;
+
+	entry->ae_par_extent.ape_epoch	= ~(0ULL);
+
+	iter_param.ip_hdl		= DAOS_HDL_INVAL;
+	iter_param.ip_ih		= entry->ae_thdl;
+	iter_param.ip_flags		= VOS_IT_RECX_VISIBLE;
+	iter_param.ip_recx.rx_idx	= PARITY_INDICATOR |
+					  (entry->ae_cur_stripe.as_stripenum *
+						entry->ae_oca->u.ec.e_len);
+	iter_param.ip_recx.rx_nr	= entry->ae_oca->u.ec.e_len;
+
+	rc = vos_iterate(&iter_param, VOS_ITER_RECX, false, &anchors,
+			 agg_recx_iter_pre_cb, NULL, entry, NULL);
+	if (rc != 0)
+		goto out;
+
+	if (entry->ae_par_extent.ape_epoch > entry->ae_cur_stripe.as_hi_epoch &&
+			entry->ae_par_extent.ape_epoch != ~(0ULL)) {
+		/* Parity newer than data; nothing to do. */
+		update_vos = false;
+		goto out;
+	}
+
+	if ((entry->ae_par_extent.ape_epoch == ~(0ULL)
+				 && agg_stripe_is_filled(entry, false)) ||
+					agg_stripe_is_filled(entry, true)) {
+		/* Replicas constitute a full stripe. */
+		rc = agg_encode_local_parity(entry);
+		goto out;
+	}
+	if (entry->ae_par_extent.ape_epoch == ~(0ULL)) {
+		update_vos = false;
+		goto out;
+	}
+	/* Parity, some later replicas, not full stripe. */
+	rc = agg_process_partial_stripe(entry);
+out:
+	if (update_vos && rc == 0)
+		rc = agg_update_vos(entry);
+		/* offload of ds_obj_update (TBD) to push remote parity */
+
+	agg_clear_extents(entry);
+	return rc;
+}
+
+/* returns the subrange of the RECX iterator's returned recx that lies within
+ * the current stripe.
+ */
+static daos_off_t
+agg_in_stripe(struct ec_agg_entry *entry, daos_recx_t *recx)
+{
+	unsigned int		len = entry->ae_oca->u.ec.e_len;
+	unsigned int		k = entry->ae_oca->u.ec.e_k;
+	daos_off_t		stripe = recx->rx_idx / (len * k);
+	daos_off_t		stripe_end = (stripe + 1) * len * k;
+
+	if (recx->rx_idx + recx->rx_nr > stripe_end)
+		return stripe_end - recx->rx_idx;
+	else
+		return recx->rx_nr;
+}
+
+/* Iterator call back sub-function for handling data extents.
+ */
+static int
+agg_data_extent(vos_iter_entry_t *entry, struct ec_agg_entry *agg_entry,
+		daos_handle_t ih, unsigned int *acts)
```

Comment:
It looks to that it's more straightforward (and efficient) to iterate parity extents first? We need only to process the replicas above the last visible parity, replicas under the last visible parity can be simply deleted. (Of course, the whole iteration should be bounded by a epoch range specified by upper caller)